### ISSUE TITLE
Public open import in parametrised module wrongly reparametrises non parametrised definition

### ISSUE THREAD

**Issue by gallais** [agda/agda#4057]
This is a problem we have encountered many times in the standard library.
I have finally managed to write a test case for it.

First, have a parametrised module re-exporting some non-parametrised definitions.

```agda
module openpublic (A : Set) where

open import Agda.Builtin.Bool public
```

Until there everything is fine: whether you import `openpublic` applied or not,
`Bool` is still a `Set` (and not a `Set -> Set`).

Then have a doubly parametrised module and use only one of the parameters to
open publically an applied `openpublic`:

```agda
module openpublic2 (A B : Set) where

open import openpublic A public
```

If you now import `openpublic2`, `Bool` has become a `Set -> Set -> Set` function:

```agda
open import openpublic2

_ : Bool ? ?     -- wtf?
_ = true {?} {?} -- wtf?
```

Ping agda/agda-stdlib#553, agda/agda-stdlib#684, agda/agda-stdlib#719

**Comment by MatthewDaggitt** [comment]
Yes!! This would be amazing to fix :+1: :+1: (Am I being too enthusiastic here?)

**Comment by nad** [comment]
For consistency, wouldn't it be better to add the telescopes to all definitions in `openpublic`, whether they are defined locally or re-exported? (See #892 for a long discussion about this, including some arguments against such a change.)

I saw that you [rely on](https://github.com/agda/agda-stdlib/blob/835a59a741eb27381f1cab946505d25775834517/src/Function.agda#L9-L25) Agda's current behaviour in the standard library, but I'm leaning towards preferring a "cleaner" design.

**Comment by gallais** [comment]
Note that it is currently impossible to `open ... public` before the module's name
(and telescope of parameters). If it were possible, we would be able to distinguish
between a reparametrised open-public and non reparametrised one.

**Comment by MatthewDaggitt** [comment]
> I saw that you rely on Agda's current behaviour in the standard library, but I'm leaning towards preferring a "cleaner" design.

Yes it's fairly deeply embedded in the standard library since at least v0.12 via `Algebra.FunctionProperties` exporting `Algebra.FunctionProperties.Core`.

My main question is that if you remove it, how can we get the same effect? It's a **really** useful feature when trying to make summary modules that re-export other modules. In fact just earlier this week I opened an [issue](https://github.com/agda/agda-stdlib/issues/876) to discuss using it to greatly standardise such "summary" modules. I for one feel that losing this behaviour would be a significant loss in functionality. 

If I understand the conclusion in #892 correctly, it was not to change it in the direction your suggesting @nad? Mostly because of backwards compatibility concerns. @gallais is suggesting changing it in the other direction which would have far less problems, and indeed I can't see anyone complaining about the new behaviour. Who needs extra irrelevant parameters attached to a definition?

**Comment by nad** [comment]
I just think that the current feature is not very aesthetic. If I have a function `f x = record { … }`, then I do not expect to be able to project out any field without supplying an argument to `f`. Why should this be possible for parametrised modules of the form `module M (x : …) where …`?

**Comment by ghost** [comment]
Here is a (perhaps more compelling) argument against this feature:

If you write `module M (x : X) where n = 17`, the `n` declaration can’t be used without an `X`, so why should a module import statement behave differently?

This seems to me like unexpected behavior, and an open gate for mistakes and accidents (such as the one described in this very issue).

As an example of another possible mistake, I would expect `import M ; f = M.f` to behave the same as `open import M using (f) public` in most (perhaps all) situations (even when inside a module with parameters).

Instead of trying to mitigate the issue by either raising a warning or, e.g., only adding the used parameters (which doesn’t necessarily work if a parameter’s type depends on another’s), I feel like it would be better to eliminate the issue through its root. (I.e. to not encourage the use of this feature.)

Personally, I feel like this feature also feels incomplete, since it requires declarations to be in a different *file* in order to be able to benefit from not having unnecessary leading parameters. I.e., single functions cannot benefit from it even if they don’t use some (or any) parameters unless they are promoted to their own files.

---

I feel like employing other Agda features would be a much more appropriate solution for the standard library. For example:

~~~Agda
module Function where

open import Function.Core public

-- Potentially give it a name to allow users to pass the arguments only once:
module _ {A B} (_≈₁_ : Rel A 0ℓ) (_≈₂_ : Rel B 0ℓ) where
	import Function.Definitions _≈₁_ _≈₂_ as Definitions
	import Function.Structures _≈₁_ _≈₂_ as Structures

open Definitions public
open Structures public
~~~

*Edit: Fixed syntax.*

(I used `0ℓ` as an example for clarity.)

Sure, we lose a bit of convenience, since we would need to add another `open` statement for each module like this, but I definitely feel like this feature is not worth the awkwardnesses it causes.

It feels compelling since, per what I said, it’s more convenient, but I don’t think it’s justified, as it can be confusing, and the boilerplate it helps avoid is not that meaningful anyway; it only saves a single line per module import that would use the feature (which I’d argue shouldn’t be many anyway).

Either way, note that I’m not sure whether it would be a good idea to take it away now that it has been introduced for so long, but I feel like at least its use should be discouraged (perhaps with a warning of sorts, or something).

**Comment by gallais** [comment]
I do not understand how your proposition is supposed to help: `Function` is not
a parametrised module anymore and as such cannot be applied upon import. We have
to first import it then `open` it `using` the non parametrised things and finally
open every single parametrised sub-module applied.

Earlier in this thread, I suggested the ability to `open (...) public` *before*
the top-level `module` declaration. It would give us the means to state that
these re-exported definitions should not be reparametrised whilst embracing the
uniform reparametrisation of everything under the top-level module declaration.

Instead of the current `Function` package in the stdlib we could then write
the following and obtain the same set of functions:

```agda
open import Function.Core public
open import Function.Packages public

open import Relation.Binary using (Rel)

module Function
  {a b ℓ₁ ℓ₂} {A : Set a} {B : Set b}
  (_≈₁_ : Rel A ℓ₁) -- Equality over the domain
  (_≈₂_ : Rel B ℓ₂) -- Equality over the codomain
  where

open import Function.Definitions _≈₁_ _≈₂_ public
open import Function.Structures  _≈₁_ _≈₂_ public
```

**Comment by ghost** [comment]
Yes, I acknowledge that, but it’s important to note what I wrote in my comment (inside the code).

In that case, instead of “opening every parameterized module applied”, one could open the explicitly‐named module once and be done with that.

~~~Agda
open import Function
open WithEqualities _==_ _==_ -- Or whatever other better name.
~~~

I know this is not a solution that reproduces the exact behavior of the feature.

Either way, what I’m saying is that I personally don’t see the “convenience” argument as too compelling (as I had said). I don’t think there are comparatively many examples of where this would cause a major boilerplate reduction (or clarity enhancement).

It’s not that I didn’t see or like your suggestion, but I’m not sure about when it should stop. If someone can `open … public` outside the top level module, then why would they not be able to declare other things such as functions? What if they want their function to use a declaration from within the top level module, perhaps in a mutually‐recursive way? Also, if they don’t need their module to have any parameter, then would the `module` declaration not be redundant?

It also *still* feels incomplete (and I don’t think would even solve the problem you are proposing in this issue’s description). What if someone wants only some of their parameters to be used in a module reexport?

~~~Agda
-- Where do we put the following line?
-- open import openpublic A public

-- Outside the module? But `A` is not available.
module openpublic2 (A B : Set) where
-- Inside the module? But an unwanted `B` parameter is also added to `Bool`.
~~~

**Comment by MatthewDaggitt** [comment]
> I know this is not a solution that reproduces the exact behavior of the feature.

> I don’t think there are comparatively many examples of where this would cause a major boilerplate reduction (or clarity enhancement).

That's the crux of this from my point of view. As far as I'm aware, this behaviour is impossible to replicate using other features of Agda. It's not a matter of reducing boilerplate. 

> It’s not that I didn’t see or like your suggestion, but I’m not sure about when it should stop. If someone can open … public outside the top level module, then why would they not be able to declare other things such as functions? What if they want their function to use a declaration from within the top level module, perhaps in a mutually‐recursive way? Also, if they don’t need their module to have any parameter, then would the module declaration not be redundant?

@gallais's proposal is simply to make this behaviour consistent. I'm not sure I quite buy your "thin end of the wedge" argument, that we shouldn't fix this inconsistency in the current implementation, because then people might want to add additional behaviour... Surely each addition can be judged on its own merit, as and when people request it?

**Comment by ghost** [comment]
Yes, there is no way to reproduce it using other features, sure. But what I meant when I talked about boilerplate is for *importing* modules, and not declaring them. That is, people would need to add an extra `open` declaration when using your module. I’m just saying that you could avoid relying on the feature by requiring library users to add such an `open`.

---

I’m saying that I think an `open` declaration should behave similarly to adding the imported declarations in its place. It just seems inconsistent to me that you would be able to have a `open` statement before the top level module, and not actual declarations.

Also, it doesn’t really seem like a much better solution from a “consistency” point of view. Why would something declared *outside* the module be considered to be inside it? Nested modules don’t behave that way at all, why should top level ones do?

(Also, to reiterate, it doesn’t solve the problem in this issue’s description at all.)

Here’s another (hopefully) compelling argument against the idea (which is a combination of some of my previous ones): What if someone wants to use a nontrivial function declaration as the argument to a module outside the top level module? They can’t at all without allowing function declarations outside the top level module too.

**Comment by gallais** [comment]
> They can’t at all without allowing function declarations outside the top level module too.

They can declare it in a separate module which they can then import before the
top-level module declaration as is already possible.

**Comment by UlfNorell** [comment]
The core issue (which is discussed at length in #892) here is the difference between module application and `open public`. This is further obscured by the sugar that lets you do both at the same time. To reiterate:
- `open M public`: name juggling that let's you access the *names* from `M` as if they were defined in the enclosing module
- `module M tel = M' es`: create new module `M` containing wrapper functions for everything from `M'`
- `open M es public` is sugar for `module Fresh = M es; open Fresh public`

The OP can be simplified to the following test case illustrating the issue:
```agda
module _ where

module openpublic (A : Set) where
  open import Agda.Builtin.Bool public

module openpublic2 = openpublic

-- openpublic.Bool is a scope alias for the name Agda.Builtin.Bool.Bool
_ : Set
_ = openpublic.Bool

-- openpublic2.Bool is a definition in the parameterised module openpublic2:
-- module openpublic2 (A : Set) where
--   Bool = openpublic.Bool
_ : Set → Set
_ = openpublic2.Bool
```

As discussed in #892 changing `open public` to introduce wrapper functions is prohibitively expensive and would mean we lose the ability to express what it currently does.

What would make sense (I believe), and what the OP suggests, would be to change it so that no wrappers are generated for `open public`'d names and instead scope alias them.

**Comment by andreasabel** [comment]
During some archeology on scope checking record definition I stumbled on the commit where the open-public evil entered the garden of Eden: df2dbefe8c42186996ca642a51043530d22f6de5 🐍 🍎 
> improved implementation of open public
> ...
> open public no longer creates new definitions for the definitions in the opened module. This has the minor effect on the semantics of open public in parameterised modules that the opened definitions no longer inherit the parameters of the module into which they are opened.

Notice the qualification as "minor effect" 😆 .
